### DISCUSSION TITLE
HyperionSPI vs HyperSerialESPx

### DISCUSSION THREAD

**Discussion by bruvv** [awawa-dev/HyperHDR#470] [Q&A]
So I am waiting for my led strip to arrive but I am trying to understand as much as possible. What I have is the following:

- Raspberry Pi 3B
- ESP8266 (D1 mini)
- SK6812 (RGBW, CW)
- Level shifter

But I am at a lost what the best and feature proof solution is to go with: [HyperionSPI](https://github.com/awawa-dev/HyperSPI) or [HyperSerialESP8266](https://github.com/awawa-dev/HyperSerialEsp8266) what is the difference anyways?

**Comment by awawa-dev** [comment]
Hi
As you can see the statistics on projects homepages ESP32 offers about 10% better refresh rate than Esp8266 for serial solutions. I don't know unfortunately HyperionSPI.

**Reply by bruvv** [reply]
Thanks for the reply, not asking about the difference in ESP32 and ESP8266 but what the essential difference are between SPI and ESP8266? Why does HyperSPI exist it seems like it is the same page as for ESP8266.

**Reply by awawa-dev** [reply]
Simply, Esp8266 is currently not best choice since there is a lot of counterfeits, doesn't allow pinout remapping, doesn't support multisegments, has worse statistics and ESP32 could be a cheaper solution. Main advantage of **HyperSPI** is that solution is not interfere with USB bus...but it's described also on the project homepage.

**Reply by bruvv** [reply]
For me I do not understand the physical difference between them. Can you make a drawing to make it visual what the difference is between them ? As I do not see how you connect the leds to the USB and power the esp at the same time.

**Comment by bruvv** [comment]
To anwser my own question:
HyperSPI = RPI > USB > ESPx
HyperESPx = RPI > GPIO > ESP GPIO

This was not clear for me. Now it make much more sense

**Reply by KookieMonster** [reply]
Hey there... could you clarify this a bit? Because I managed to get my setup working by flashing HyperSPI to the ESP32 but the connections are a little messy. Here's what I have and how I made it work:

Rpi 3B+ (running HyperHDR 19.1)
ESP32 (DOIT DEVKITV1) (running HyperSPI NeoPixels version)
Level Shifter (SN74AHCT125N)
SK6812 RGBW NW (yeah, should have gone with Cold, I know that now)

**Power Supply -> ESP32**
V+ -> VIN
V- -> GND
**ESP32 to Rpi**
USB cable
GPIO 5 -> SPI Chip Enable
GPIO 18 -> SCK
GPIO 19 -> MISO
GPIO 23 -> MOSI

So... my question is, is it really necessary to have the ESP32 powered by both VIN+GND and USB Cable to the Rpi? And how different would the connections be if I flashed HyperSerialESP32?
Also... what would be the best option for my setup? HyperSPI or HyperSerialESP32?

Thanks in advance!!

**Reply by KookieMonster** [reply]
Hey @awawa-dev , sorry to bother you mate, but would you care to enlighten me on these doubts of mine? 😄 

My ESP32 is packed with CH9102x chip. Forgot to mention that.

Thanks in advance.

**Reply by bruvv** [reply]
No the ESP is powered by the USB of the RPI.

**Reply by KookieMonster** [reply]
> No the ESP is powered by the USB of the RPI.

Thanks for the reply mate.

The thing is... if I unplug the 5V from the VIN, the strip stops taking data. I wonder why...?

**Reply by bruvv** [reply]
because you connected both, replug the usb and it should work.

**Reply by KookieMonster** [reply]
Uh... that makes sense! Thanks mate!

> And how different would the connections be if I flashed HyperSerialESP32?
> Also... what would be the best option for my setup? HyperSPI or HyperSerialESP32?

Still need to clarify this though...

**Comment by bruvv** [comment]
I never fully understood it but as far as I understand: HyperionSPI = raspberry SPI en HyperSerialESP32 = SPI from the ESP32.

**Comment by awawa-dev** [comment]
Really, all can be found in the projects specification:
HyperSPI: better latency thanks to SPI speeds, doesn't suffer from Rpi USB2.0 limited USB bus, can be use at short distance (generic SPI characteristics), doesn't required serial chip on ESP board that is 2Mb capable
HyperSerialESP32: easier to setup, can be used at longer distance (generic USB benefit), may can cause USB2.0 grabber to reduce performance due to USB2.0 Rpi limited bus, a bit worse latency but recommended boards can work at 4Mb or even 5Mb speed.

And @bruvv : there is no HyperionSPI project.

**Reply by KookieMonster** [reply]
Yeah, I know mate. I've been reading it all but just had a hard time grasping a few things. But anyway... thanks for the explanation, the great work and awesome support.

I think I'll stick with HyperSPI then.

**Reply by bruvv** [reply]
> And @bruvv : there is no HyperionSPI project.

Hu? What is this ? https://github.com/awawa-dev/HyperSPI or do you mean my typo? I meant hyperSPI instead of hyperionSPI
